Span Processing Pipelines (#9966)
DataDog/documentation: content/en/tracing/processing_pipelines/_index.md

---
title: Span Processing Pipelines
kind: documentation
description: 'Enrich span metadata with processing pipelines.'
further_reading:
    - link: 'tracing/trace_retention_and_ingestion'
      tag: "Documentation"
      text: 'Customize trace ingestion and retain important traces.'
    - link: 'tracing/trace_search_and_analytics/query_syntax'
      tag: "Documentation"
      text: 'Use Analytics queries and monitors based on retained traces.'
---
<div class="alert alert-warning">

> This feature is currently in private beta. <a href="https://docs.google.com/forms/d/1RlT0FNdFjiEzkQgxiCf77ugpW0w5a17X7JQ7E286jM4">Fill out this form</a> to request this feature for your account or to be notified when it becomes generally available.

Author: ```suggestion
This feature is currently in private beta. <a href="https://forms.gle/6hHcytp2JvBAHxC6A">Fill out this form</a> to request this feature for your account or to be notified when it becomes generally available.
```

### Pipeline filters

Filters let you define and limit what kinds of spans a pipeline applies to. The syntax is the same as when writing a [search query][7] or [retention filter][8].

**Note:** Filtering is applied before any of the pipeline’s processors, so you cannot filter on an attribute that is yet-to-be-created in the pipeline.

### Nested pipelines

_Nested pipelines_ are pipelines created within a pipeline. Use nested pipelines to split the processing into separate steps. For example, first use a high-level filtering such as `env` and add common processors, and then a second level of filtering based on `service`, or any other tag or attribute and their custom processors.

**Note:** You can nest pipelines only one level deep. Within a nested pipeline, you can create processors, but not additional nested pipelines.

## Processors

A processor executes within a pipeline to complete a data-structuring action, such as arithmetic computation or tag enrichment, on a trace span.

### Arithmetic processor

{{< img src="tracing/processing_pipelines/arithmeticprocessor.png" style="width:100%;" alt="How to create an arithmetic processor" >}}

With arithmetic processors, you can:

> - Add new attributes to a span with the result of a formula that you define. Added attribute names must not contain spaces or special characters.

Reviewer: ```suggestion
- Add new attributes to a span with values that are the calculated result of a formula that you define. Added attribute names must not contain spaces or special characters.
```

- Remap different time attributes with different units into a single attribute.
- Compute operations on attributes within the same span.

Arithmetic processors follow these rules:

1. Formulas can use parentheses and the basic arithmetic operators: `-`, `+`, `*`, and `/`.

> 2. Attributes must be prefixed with `@` when used in formulas.

Reviewer: ```suggestion
2. In the formula, prefix attributes names with `@`.
```

> 3. By default, the calculation is skipped if an attribute is not present on the span, or if it cannot be parsed as a number.
> 4. If desired, select the option **Replace missing attribute by 0** to automatically populate missing attribute values with `0` to ensure that the calculation is done.

Reviewer: ```suggestion
3. By default, if the formula references an attribute that is not present on a span, or if it cannot be parsed as a number, the processor is skipped for the span. However, you can change this behavior by selecting **Replace missing attribute by 0**, and the formula will used `0` for any missing attribute values, to ensure that the calculation is done.
```

> 5. If you use `-` as an operator, it must be used with a space on either side like `a - b`, as `-` can also exist in attribute names.

Reviewer: ```suggestion
5. If you use `-` as an operator, put a space on either side, as in `a - b`, because `-` can also exist in attribute names.
```

> 6. If the target attribute already exists, it is overwritten by the result of the formula.

Reviewer: ```suggestion
6. If the target attribute already exists, its value is overwritten by the result of the formula.
```

7. Results are rounded up to the ninth decimal place. For example, if the result of the formula is `0.1234567891`, the actual value stored for the attribute is `0.123456789`.

### Lookup Processor

{{< img src="tracing/processing_pipelines/lookupprocessor.png" style="width:100%;" alt="How to create a lookup processor" >}}

Lookup processors allow you to define a mapping between a span attribute and human readable values saved in an [Enrichment Table](#enrichment-tables).  The Enrichment Tables feature is also in beta.

For example, with a lookup processor you can:
- Map an internal service ID into a human readable service name.
- Store an ID value associated with address and other personal data that is either subject to frequent changes or shouldn't be set within application code.
- Check if the MAC address that just attempted to connect to the production environment belongs to your list of stolen machines.

A lookup processor performs the following actions:

1. Looks to see if the span contains the source attribute.

> 2. Checks if the source attribute value exists in the mapping table.
> 3. If it does, creates the target attribute with the corresponding value from the table.
> 5. If the target attribute already exists, it is overwritten by the result of the formula.

Reviewer: ```suggestion
2. Checks if the source attribute value exists in the mapping table, and if it does, either:
    - Creates the a span attribute with the corresponding value from the table; or
    - Overwrites the span's existing attribute with the value from the table.
```

> 6. Optionally, if a lookup processor does not find the source attribute value in the mapping table, creates a target attribute with the fixed default value.

Reviewer: How do they specify the fixed default value?
